### PR TITLE
[blend2d] update port to 2024-06-28

### PR DESCRIPTION
**Pull request by JakubMelka**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Review by GO-NFT-GO — COMMENTED** [review]
Ok

**Comment by dg0yt** [comment]
Shouldn't this switch to `BLEND2D_EXTERNAL_ASMJIT`?

**Comment by JakubMelka** [comment]
To @dg0yt : I think this is intentional - to make Blend2d standalone library without dependencies. I am not author of Blend2d, @kobalicek is the author, maybe he can explain this? I did not like to change it in this pull request.

**Comment by dg0yt** [comment]
At least in this case, there is already a written guideline:
Do not use vendored dependencies
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#do-not-use-vendored-dependencies

**Comment by kobalicek** [comment]
Switching to `BLEND2D_EXTERNAL_ASMJIT` is the way to go for conan and vcpkg - it was added there for this exact purpose so the package manager can resolve the dependencies.

However, of course this means asmjit is not configured by Blend2D, which means a larger resulting binary as asmjit by default compiles ALL backends in, and not just the native one.

**Comment by kobalicek** [comment]
I think additional question is whether not to switch to the release numbers of Blend2D. I can even make a new patch release so this port gets the latest code.

**Comment by JakubMelka** [comment]
OK, @kobalicek, I will try to switch to external dependency. However, arm64_windows compilation fails and I do not know why. How to solve it?

**Comment by kobalicek** [comment]
Is there a way to see what failed in arm64_windows case?

I think this is super confusing. You see it failed, but where is the build log if build failed?

**Comment by JakubMelka** [comment]
@kobalicek, logs are available and I am attaching them. 
[failure logs for arm64-windows.zip](https://github.com/user-attachments/files/16116923/failure.logs.for.arm64-windows.zip)

**Comment by JakubMelka** [comment]
@kobalicek, it seems that there is a problem in this file:
opentype/otglyfsimdimpl_p.h(35): error C2871: 'SIMD': a namespace with this name does not exist

But I think only in DEBUG configuration.

**Comment by kobalicek** [comment]
I will get this fixed upstream, it's a pretty trivial problem.

**Comment by JakubMelka** [comment]
@dg0yt, I have switched to BLEND2D_EXTERNAL_ASMJIT, asmjit is now external dependency.

**Comment by kobalicek** [comment]
I have released a Blend2D 0.11.3 with the fix regarding Windows/ARM64 build.

**Comment by kobalicek** [comment]
And regarding the build errors.

I think I know the problem. Blend2D only exports C-API so it's compiled with `-fno-rtti` and `-fno-exceptions`. However, if you consume AsmJit that is compiled with exceptions and RTTI, that can create a problem I think.

And yeah - this is the main reason it's bundled, so it's not compiled with different compilation flags. And I'm not sure how to solve it, just please don't bloat Blend2D with RTTI or Exceptions.

**Comment by dg0yt** [comment]
> I think I know the problem. Blend2D only exports C-API so it's compiled with `-fno-rtti` and `-fno-exceptions`. However, if you consume AsmJit that is compiled with exceptions and RTTI, that can create a problem I think.

This problem shouldn't exist with CMake config, by passing through transitive usage requirements such as linker language.

Looking at the logs:

~~~
CMake Error at /mnt/vcpkg-ci/installed/x64-linux/share/blend2d/blend2d-config.cmake:60 (set_target_properties):
  The link interface of target "blend2d::blend2d" contains:

    asmjit::asmjit
~~~

`blend2d` CMake config seems to need a `find_dependency`.

**Comment by kobalicek** [comment]
I have some additional remarks regarding this port:

  - Description should be "2D Vector Graphics Engine Powered by a JIT Compiler"
  - Supports: "!wasm32" - not sure why UWP should not be supported, maybe there was a reason for that in the past, but I would at least verify that.
  - When looking into `vcpkg.json` it only enables JIT on Windows platform. This is wrong as JIT is supported on vcpkg supported platforms that target either X86, X86_64, or AArch64 architectures. This should be fixed.
  - The feature `logging` should be removed. It's mostly for development or for shrinking few kB of final binary, not worth having a flag for it.
  - The feature `tls` is only enabled on Windows, which is wrong. This should be removed as well as there is no platform supported by vcpkg where TLS would not work. This is again mostly for development and CI to ensure Blend2D builds on platforms without TLS.

This should be corrected as otherwise this port is providing a crippled Blend2D without a working JIT on non-windows platforms, but JIT is essential for this library. This is basically against upstream.

Regarding AsmJit dependency - I think there must be a problem with vcpkg fixup, because AsmJit is a PRIVATE dependency of Blend2D, so if it gets into INTERFACE it must be done in some step that is not done by upstream.

**Comment by dg0yt** [comment]
> No - Blend2D config doesn't need a find_dependency(), because AsmJit is internal.

AFAIU `BLEND2D_EXTERNAL_ASMJIT` means that it is not internal. And it is not difficult to configure the exported CMake config from 
~~~
if("@BLEND2D_EXTERNAL_ASMJIT")
    ...
endif()
~~~

And add a `@BUILD_SHARED_LIBS@` condition if `asmjit::asmjit` is used with `PRIVATE`.

> In retrospective, I think BLEND2D_EXTERNAL_ASMJIT was a mistake as it gets more problematic than just having a single internal dependency. If I cannot tell vcpkg how to compile it properly, then it makes no sense to expose it as a dependency.

There are really more complex compositions, and usually it works well with CMake config (at least until requiring *incompatible* configurations).

**Comment by JakubMelka** [comment]
@dg0yt, could you please help me? I do not know exactly, what to change? Where to place this code?

**Comment by dg0yt** [comment]
Okay, I will have a look.

**Comment by JakubMelka** [comment]
Thank you.

**Comment by dg0yt** [comment]
FTR: This is from x64-linux (static library linkage):
~~~
set_target_properties(blend2d::blend2d PROPERTIES
  INTERFACE_COMPILE_FEATURES "cxx_std_11"
  INTERFACE_COMPILE_OPTIONS "-DBL_STATIC"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include;${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "\$<LINK_ONLY:c>;\$<LINK_ONLY:pthread>;\$<LINK_ONLY:m>;\$<LINK_ONLY:asmjit::asmjit>"
)
~~~
And that's why blend2d cmake config must do `find_dependency` for static linkage. 
I will try to prepare a change which could be integrated upstream.

**Comment by dg0yt** [comment]
> * not sure why UWP should not be supported, maybe there was a reason for that in the past, but I would at least verify that.

~~~
FAILED: CMakeFiles/blend2d.dir/src/blend2d/filesystem.cpp.obj 
C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1438~1.331\bin\Hostx64\x64\cl.exe   /TP  -external:ID:\installed\x64-uwp\include -external:W0 /DWIN32 /D_WINDOWS /D_UNICODE /DUNICODE /DWINAPI_FAMILY=WINAPI_FAMILY_APP /D__WRL_NO_DEFAULT_LIB__ /nologo /Z7 /MP /GS /Gd /Gm- /W3 /WX- /Zc:wchar_t /Zc:inline /Zc:forScope /fp:precise /Oy- /EHsc  /utf-8  /D_DEBUG /MDd /Od /RTC1  -MDd -MP -GR- -GF -Zc:__cplusplus -Zc:inline -Zc:strictStrings -Zc:threadSafeInit- -W4 -DBL_BUILD_OPT_AVX512 -GS /showIncludes /FoCMakeFiles\blend2d.dir\src\blend2d\filesystem.cpp.obj /FdCMakeFiles\blend2d.dir\ /FS -c D:\b\blend2d\src\35b6a6f31f-2f74679bb4.clean\src\blend2d\filesystem.cpp
cl : Command line warning D9025 : overriding '/W3' with '/W4'
D:\b\blend2d\src\35b6a6f31f-2f74679bb4.clean\src\blend2d\filesystem.cpp(289): error C3861: 'CreateFileW': identifier not found
D:\b\blend2d\src\35b6a6f31f-2f74679bb4.clean\src\blend2d\filesystem.cpp(853): error C3861: 'CreateFileMappingW': identifier not found
D:\b\blend2d\src\35b6a6f31f-2f74679bb4.clean\src\blend2d\filesystem.cpp(857): error C3861: 'MapViewOfFile': identifier not found
~~~
These Windows APIs aren't available for UWP.

**Comment by JakubMelka** [comment]
I will disable uwp until upstream solves this.

**Comment by dg0yt** [comment]
If hold your branch stable for a moment, I create a ready-to-use PR in your repo.

**Comment by JakubMelka** [comment]
@dg0yt, OK, I will not commit anything until your pull request.

**Comment by dg0yt** [comment]
https://github.com/JakubMelka/vcpkg/pull/1
https://github.com/blend2d/blend2d/pull/207

**Comment by JakubMelka** [comment]
Thank you, @dg0yt . I will wait for @kobalicek to merge it to the master branch and then I will update this pull request.

**Comment by dg0yt** [comment]
I would suggest to validate it here already. I could only test with x64-linux locally.

**Comment by JakubMelka** [comment]
How should I do that, if it is not merged?

**Comment by dg0yt** [comment]
You just merge my PR into your branch. It is just pressing the Merge button in your GH repo.
(I am a community member. I have no privileges to write to your PR directly.

**Comment by JakubMelka** [comment]
I think @kobalicek is only who can do that, because I have no write access to his repository.

**Comment by dg0yt** [comment]
I'm talking about https://github.com/JakubMelka/vcpkg/pull/1.

**Comment by JakubMelka** [comment]
Oh, I overlooked this. I have merged it now.

**Comment by JakubMelka** [comment]
To @kobalicek
- Description was changed to "2D Vector Graphics Engine Powered by a JIT Compiler"
- UWP is not supported, because Blend2d uses some WinApi functions not available on UWP
- JIT is now enabled on all platforms
- Logging option was removed
- Tls is now enabled on all platforms
- Asmjit is now as external dependency

**Comment by JakubMelka** [comment]
@dg0yt, thank you for patch. It seems to be working.

**Review by Cheney-W — CHANGES_REQUESTED** [review]

File: ports/osgearth/portfile.cmake [hunk 1]
```diff
@@ -11,6 +11,7 @@ vcpkg_from_github(
 		remove-lerc-gltf.patch
 		export-plugins.patch
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply ports/osgearth/portfile.cmake:12]: Read: These lines should use spaces, not tabs.

**Comment by kobalicek** [comment]
Upstream changes: UWP should be supported now, @dg0yt PR has been merged.

It would be great to test the current Blend2D master branch, and if it works, I would like to make a new release and then point this to that. This means that changing a version from date to a release string would be great.

**Comment by dg0yt** [comment]
> This means that changing a version from date to a release string would be great.

Can you make git tags for releases? It would be more obvious when looking at Github.
(IIUC the only current reference is "Commit Hashes" on https://blend2d.com/download.html.)

**Comment by kobalicek** [comment]
@dg0yt I'm going to use git tags after 1.0. Vcpkg doesn't consider them safe anyway, so even if Blend2D uses git tags, they would not be used by package managers.

File: ports/osgearth/portfile.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (JakubMelka) [reply ports/osgearth/portfile.cmake]: I have replaced tabs by spaces.

File: versions/b-/blend2d.json [hunk 2]
```diff
@@ -1,5 +1,15 @@
...
+    {
+      "git-tree": "dfd35fa94cf5403e7b6d0e8ae382bdb41039d4f5",
+      "version-date": "2024-07-07",
+      "port-version": 0
+    },
+    {
+      "git-tree": "85399e3584f0a1057c5d4102fead6c62f88cd08c",
+      "version-date": "2024-06-28",
+      "port-version": 0
+    },
```
[the remark this answers is not in the document]
Comment (JakubMelka) [reply versions/b-/blend2d.json]: OK, I have removed it.

**Comment by JakubMelka** [comment]
I think it will be better to solve uwp problems later and not in this PR.

**Comment by kobalicek** [comment]
> I think it will be better to solve uwp problems later and not in this PR.

It does, but I don't know how the CI sets it up. When built with `_WIN32_WINNT=0x0A00` it works, but I have no idea how it's setup on CI. I though that this would be set by the SDK if not already defined. Maybe Microsoft would tell use how to detect UWP build before even including `<windows.h>`? Any compile-time macro the compiler sets?

I don't have a Windows machine atm, so I can only guess.

**Comment by JakubMelka** [comment]
@kobalicek UWP build defines this:

`/DWINAPI_FAMILY=WINAPI_FAMILY_APP `

So, you can use:

```
#if defined(WINAPI_FAMILY) && (WINAPI_FAMILY == WINAPI_FAMILY_APP)
    // UWP build
#else
    // Standard windows build
#endif
```

**Comment by JakubMelka** [comment]
@kobalicek, I have temporarily disabled file access function from UWP platform, until it is resolved. On CI UWP platform, the functions which you have been using, are not available.

**Comment by kobalicek** [comment]
It's available, it's just about a proper detection and _WIN32_WINNT definition before including anything from the Windows SDK.

Just testing this one:

  - https://github.com/blend2d/blend2d/commit/94b243e40df6c310e8efabd19aef7fad098a8059

**Comment by JakubMelka** [comment]
@kobalicek, thank you, I tried your code in the patch and it seems to be working.

**Comment by kobalicek** [comment]
@JakubMelka Great!

BTW you can remove the upw_patch now as it's part of the master branch now. I prefer upstream fixing stuff instead of patching as patching makes further contribution always more annoying :)

One more thing I would like to verify is the `vcpkg.json` file. There are some things I found very confusing so I would like to clarify first whether they are correct. The bits from `vcpkg.json`:

```json
  "dependencies": [
    "asmjit",
    {
      "name": "vcpkg-cmake",
      "host": true
    },
    {
      "name": "vcpkg-cmake-config",
      "host": true
    }
  ],
  "default-features": [
    "jit",
    "tls"
  ],
  "features": {
    "jit": {
      "description": "Default feature. Enables jit pipeline compilation.",
      "dependencies": [
        {
          "name": "blend2d",
          "default-features": false
        }
      ]
    },
    "tls": {
      "description": "Default feature. Enables use of thread_local feature. Disable for platforms where thread local storage is expensive or not supported.",
      "dependencies": [
        {
          "name": "blend2d",
          "default-features": false
        }
      ]
    }
  }
```

I'm not sure why "jit" has "blend2d" in "dependencies". I think there should be "asmjit" - then the question is should "asmjit" be specified in root "dependencies" as well? I think no... But I think vcpkg maintainers should help here...

Also, please let's remove "tls" if we are changing this - I don't think anyone would want to turn this off anyway.

I think the bits from `vcpkg.json` should be changed to this instead:

```json
  "dependencies": [
    {
      "name": "vcpkg-cmake",
      "host": true
    },
    {
      "name": "vcpkg-cmake-config",
      "host": true
    }
  ],
  "default-features": [
    "jit"
  ],
  "features": {
    "jit": {
      "description": "Enables JIT compiler to generate optimized pipelines.",
      "supports": "!arm32",
      "dependencies": [
        "asmjit"
      ]
    }
  }
```

At least this would make sense to me. AsmJit is not required when "jit" is disabled.

**Review by Cheney-W — DISMISSED** [review]

**Comment by kobalicek** [comment]
@Cheney-W There are still 2 things to do, makes no sense to approve/merge yet.

As I described - the patch is no longer necessary as it's already done by upstream, and vcpkg.json should be fixed.

**Comment by kobalicek** [comment]
@JakubMelka Are you going to fix the remaining things? I would do it myself, but I would have to do PR to your branch, I think it should be final.

**Comment by JakubMelka** [comment]
@kobalicek, Yes, I have updated the pull request. Sorry for late update, I was busy yesterday.

File: ports/blend2d/vcpkg.json [hunk 3]
```diff
@@ -18,41 +17,14 @@
     }
   ],
   "default-features": [
-    "jit",
-    "logging",
-    "tls"
+    "jit"
```
Comment (dg0yt) [inline ports/blend2d/vcpkg.json:20]: To fix default installations on arm32:
```suggestion
    {
      "name": "jit",
      "platform": "!arm32"
    }
```

**Review by dg0yt — CHANGES_REQUESTED** [review]
:+1:

**Comment by dg0yt** [comment]
`blend2d:arm-neon-android: SUCCEEDED: 24 s` :rocket:

**Comment by kobalicek** [comment]
This looks solid to me. Thanks everyone for reaching this point!

**Review by Cheney-W — APPROVED** [review]

**Pull request merged**